### PR TITLE
docs automation for rook

### PR DESCRIPTION
**Pull request by ilovett**

### CODE REVIEW

File: build/scripts/preprocess.js [hunk 1]
```diff
@@ -11,13 +11,17 @@ const semverSort = require('semver-sort');
...
 // collect all docs projects
 const data = [];
-const projects = getDirectories(`${ __dirname }/docs`);
+const projects = getDirectories(`${ ROOT_DIR }/docs`);
+
 projects.forEach(project => {
 
-  const channels = ['latest', 'stable', 'beta', 'alpha'];
-  const unsorted = getDirectories(`${ __dirname }/docs/${ project }`);
+  const channels = ['master', 'stable', 'beta', 'alpha'];
```
Comment (bassam) [inline build/scripts/preprocess.js]: why are those needed anymore?

File: build/publish.sh [hunk 2]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash -e
```
Comment (bassam) [inline build/publish.sh]: would be good to add a README file to the build/ directory to explain the approach

File: build/build.sh [hunk 3]
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash -e
+
+HOST_DIR=$(cd $(dirname "$0")/../; pwd)
+CONTAINER_DIR="/opt/rook.github.io"
+
+# install node dependencies through container
+echo "pulling npm dependencies..."
+docker run --rm --user 1000 -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine npm install
```
Comment (bassam) [inline build/build.sh]: use `--user $(id -u):$(id -g)` instead

Comment (bassam) [inline build/build.sh]: the --rm will remove the container after it exists so I'm not sure how `npm install` persist for the next run below. might be better to create a preprocess.sh that does npm install and runs node in a single script

File: build/build.sh [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+
+# run prepare script through container
+echo "preparing static assets..."
+docker run -t --user 1000 -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine node build/scripts/preprocess.js
```
Comment (bassam) [inline build/build.sh]: use `-i` instead of `-t` to avoid tty issues on concurrent builds in jenkins

File: build/publish.sh [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+HOST_DIR=$(cd $(dirname "$0")/../; pwd)
+CONTAINER_DIR="/opt/rook.github.io"
+
+# check if there are any changes to commit
+if git diff-index --quiet HEAD --; then
+  echo "no changes detected"
+else
+  echo "committing changes..."
+  git -C "$ROOT_DIR" add .
+  git -C "$ROOT_DIR" -c user.email="info@rook.io" -c user.name="Rook" commit --message="docs snapshot for rook version `$DOCS_VERSION`"
```
Comment (bassam) [inline build/publish.sh]: should this call snapshot.sh?

File: build/scripts/preprocess.js (hunk 1 above)
Comment (ilovett) [inline build/scripts/preprocess.js]: sorting bumps those channels to the top

File: build/build.sh (hunk 4 above)
Comment (ilovett) [inline build/build.sh]: I originally had `-it` and ran into the TTY error running on jenkins and found i could remove `-i` but perhaps thats backwards/wrong?

File: build/publish.sh (hunk 5 above)
Comment (ilovett) [inline build/publish.sh]: i eliminated it an split it into build/publish

File: build/scripts/preprocess.js (hunk 1 above)
Comment (bassam) [inline build/scripts/preprocess.js]: I don't think we want channels these in the list of version. just `master` and every release semver with `major.minor` only.

Comment (ilovett) [inline build/scripts/preprocess.js]: if we were going to add docs for stable/beta/alpha in the future then they would be bumped to the top of the sorting, it will be ignored if there is no folder

File: build/build.sh (hunk 3 above)
Comment (ilovett) [inline build/build.sh]: user mappings updated

File: build/build.sh [hunk 6]
```diff
@@ -0,0 +1,14 @@
...
+HOST_DIR=$(cd $(dirname "$0")/../; pwd)
+CONTAINER_DIR="/opt/rook.github.io"
+
+# install node dependencies through container
+echo "pulling npm dependencies..."
+docker run -t --user 1000 -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine npm install
+
+# run prepare script through container
+echo "preprocessing static assets..."
+docker run -t --user 1000 -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine node build/scripts/preprocess.js
```
Comment (ilovett) [inline build/build.sh]: reverted back to `--user 1000` because plays well with mac user as well as ubuntu

Comment (jbw976) [inline build/build.sh]: wouldn't that only work on your machine? i don't think I have a user 1000 on my machine.

File: build/publish.sh (hunk 2 above)
Comment (ilovett) [inline build/publish.sh]: added a `README`

File: build/build.sh (hunk 6 above)
Comment (ilovett) [inline build/build.sh]: I think it must be a fluke.  Perhaps the mapping only occurs if the given id is legit and maps to someone.  Also, running the command without user mapping at all worked on mac, but did not on ubuntu originally.

Open to suggestions on how to handle this locally on mac with the correct `--user $(id -u):$(id -g)` -- perhaps only pass this flag if non-mac ?

Comment (bassam) [inline build/build.sh]: are the permissions needed so that the the /build/scripts/preprocess.js command can execute? I wonder if you need to just run as root and chmod -R in an entry point script.

File: build/build.sh [hunk 7]
```diff
@@ -0,0 +1,14 @@
+#!/bin/bash -e
+
+HOST_DIR=$(cd $(dirname "$0")/../; pwd)
+CONTAINER_DIR="/opt/rook.github.io"
+
+# install node dependencies through container
+echo "pulling npm dependencies..."
+docker run -t --user 1000 -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine npm install
```
Comment (bassam) [inline build/build.sh]: I might be missing something. docker run start a new container, and the subsequent docker run below starts a different container. I think you are trying to run two commands in a single container, right? you might need a wrapper script to do the npm install and node build/.

Maybe this is working now because node:8-alpine already as `nom install` done?

File: build/publish.sh [hunk 8]
```diff
@@ -0,0 +1,16 @@
+#!/bin/bash -e
+
+ROOT_DIR=$(cd $(dirname "$0")/../; pwd)
+
+git -C "$ROOT_DIR" add .
```
Comment (bassam) [inline build/publish.sh]: should this be `git add -A`?

File: build/scripts/preprocess.js (hunk 1 above)
Comment (bassam) [inline build/scripts/preprocess.js]: so let's add this in the future if we need it. right now seems confusing to see channels here.

File: build/build.sh (hunk 6 above)
Comment (ilovett) [inline build/build.sh]: The problem is that the dependencies are installed with `root:root` on the ubuntu jenkins box and then the workspace cannot be removed.

File: build/build.sh (hunk 7 above)
Comment (ilovett) [inline build/build.sh]: The `package.json` in the repo root `rook.github.io` contains the definitions for which dependencies to pull, the entire repo is linked to `/opt/rook.github.io`.. running `npm install` looks at `package.json` and then pulls the code to `node_modules` in the root dir, so the next time the container is run the dependencies are present and mapped.

I had run into issues chaining commands with `docker run ... npm install && node prepare.js`

File: build/build.sh (hunk 6 above)
Comment (ilovett) [inline build/build.sh]: Perhaps the best way to handle this is to let the user mappings be environment variables and only pass the flag if they are defined giving control to the host.

File: build/build.sh (hunk 7 above)
Comment (ilovett) [inline build/build.sh]: The container essentially wraps the repo with a docker node environment so that it works identical on a machine with node installed / without docker.

File: build/build.sh [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash -e
+
+HOST_DIR=$(cd $(dirname "$0")/../; pwd)
+CONTAINER_DIR="/opt/rook.github.io"
+
+if [ -n "$HOST_USER" ]; then
+  DOCKER_RUN="docker run -t --user $HOST_USER:$HOST_USER -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine"
+else
+  DOCKER_RUN="docker run -t -w $CONTAINER_DIR -v $HOST_DIR:$CONTAINER_DIR node:8-alpine"
+fi
```
Comment (ilovett) [inline build/build.sh]: this gives flexibility to the runner / host to specify the mappings if any.  on mac, not specifying a user, the `node_modules` dependencies will be installed with current user permissions, however on linux / jenkins build, the `node_modules` dependencies maintain the `root:root` permissions from how the container is run.  then wiping the folder on the jenkins box becomes and issue due to root permissions.  this gives flexibility to the runner to specify the `$HOST_USER` user id.

File: build/publish.sh (hunk 2 above)
Comment (ilovett) [inline build/publish.sh]: please resolve this discussion thread here

File: build/build.sh [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+
+# install node dependencies through container
+echo "pulling npm dependencies..."
+$DOCKER_RUN npm install
```
Comment (bassam) [inline build/build.sh]: does `npm install` only install files in the working directory? Is this way two instances of `docker run` succeed?

Comment (ilovett) [inline build/build.sh]: yes, for node projects dependencies are pulled into `${projectDir}/node_modules` and relative to `package.json`
